File: services/rust/poem/src/routes/handlers/get.rs
use crate::routes::{errors::AppError, extractors::AppPath};
use app::state::AppState;
use poem::{
    Response,
    error::ResponseError,
    web::{Data, IntoResponse, Json},
};
use std::sync::{Arc, Mutex};
use tracing::{error, info, instrument, warn};

#[poem::handler]
#[instrument(skip_all)]
pub async fn get_handler(
    Data(state): Data<&Arc<Mutex<AppState>>>,
    AppPath(id): AppPath,
) -> Response {
    if let Ok(state) = state.lock() {
        if let Some(task) = state.tasks.get(&id).cloned() {
            drop(state);
            info!(
                %id,
                secret = task.secret.len(),
                operation = task.operation.to_string(),
                method = "GET",
                "Retrieved task"
            );

            return Json(task).into_response();
        } else {
            drop(state);
            warn!(%id, method = "GET", "Task not found");
            return AppError::TaskNotFound.as_response();
        }
    }

    error!(%id, method = "GET", "Poisoned lock");

    AppError::Internal.as_response()
}

File: services/go/echo/routes/handlers/get.go
package handlers

import (
	"app"
	"echo/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/labstack/echo/v5"
)

func GetHandler(state *app.AppState) echo.HandlerFunc {
	return func(c *echo.Context) error {
		id, err := ParseID(c)
		if err != nil {
			return err
		}

		state.Mu.RLock()
		task, ok := state.Tasks[id]
		state.Mu.RUnlock()

		if !ok {
			slog.Warn("Task not found", "method", c.Request().Method, "path", c.Request().URL.Path, "id", id.String())
			return routes.AppErrors.TaskNotFound.Error(c)
		}

		slog.Info(
			"Retrieved task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", c.Request().Method,
		)

		return c.JSON(http.StatusOK, task)
	}
}
